### DISCUSSION TITLE
How to set up integral in electromagnetic far-field calculation?

### DISCUSSION THREAD

**Discussion by M8kmyday** [mfem/mfem#4516] [Q&A]
I have the $$\overline{E}$$ and $$\overline{H}$$ solution on an outer box where a 1st-order radiation boundary has been applied.  I want to calculate the far-field from $$\overline{E}$$ and $$\overline{H}$$ to get the antenna radiation pattern.  In that calculation, an integral appears of the form of $$\iint_S (\hat{n}\times\overline{H})e^{jk_o \overline{r}^{\prime}\cdot\hat{r}}dS$$, where $$\overline{r}^{\prime}$$ ranges over the radiation boundary and $$\hat{r}$$ is a point in the far field.  The integration has to be repeated for each point in the far field to fill out a far-field plot.

I think I need a Coefficient that is spatially dependent on the boundary ($$\overline{r}^{\prime}$$) and that takes a coordinate ($$\hat{r}$$) as a parameter.  Is that the best way to approach this integral?  If so, I have reviewed the available Coefficients, and I do not see one that fits the bill, meaning a custom Coefficient is required.  Is that correct?  I think I can create a custom Coefficient, but if there is a better way, I would prefer not to.

**Comment by M8kmyday** [comment]
Can someone spare some time to comment on this?  It would be greatly appreciated.

**Reply by tzanio** [reply]
ping: @mlstowell, @psocratis

**Reply by M8kmyday** [reply]
I don't think that a Coefficient is the way to go.  I'm seeing a few possibilities:
1. Extract an x-y grid of points from the solution than numerically integrate.  This would be slow and inelegent.
2. Fit the exponential function similar to fitting done in Example 38, then integrate using the LinearForm infrastructure.  This would also be slow since the fitting would have to be repeated for every $$\hat{r}$$.
3. Extract the polynomial fit for each boundary element, then calculate the integral using custom formulas for the polynomial terms multiplied by the exponential.  This would be fast, but most of the calculation would take place outside of MFEM.
4. Put in place custom quadradure formulas calculated with the exponential applied, then integrate in MFEM.  This is very similar to 3.

**Comment by mlstowell** [comment]
Hi, @M8kmyday,

This is not something I have much experience with but I did once have a colleague who was an expert in such things and I picked up a basic understanding. The short answer is that this is a computationally expensive endeavor unless the far field point is very far from the computational domain. The computation is what you label as possibility 2 i.e. compute a LinearForm for each far field point. The inner product of the Linear Form with your H field vector gives the far field value (scalar values only, if you need vector values you will need a Linear form for each component.

Clearly this is quite expensive if you have many far field points. To speed this up people have developed various means of approximating these values by breaking up the boundary into smaller pieces and computing linear forms over the individual pieces rather than the entire boundary. The advantage of this is that some of these pieces might be farther from a given evaluation point so the integral over these farthest pieces can be replaced by single point integrations based on field data at the center of such a piece and its distance, orientation, and size.

You are probably just as familiar as I am with these methods. They typically go by names such as the "method of moments" (which is not precisely what I'm describing) or "multilevel fast multipole method". Unfortunately, MFEM does not have much of the infrastructure needed to perform such calculations.

Best wishes,
Mark

**Comment by M8kmyday** [comment]
Thanks for the background and detail.  It seems that there isn't an MFEM-based method that I missed that would have been a silver bullet to make the calculation.  I'm in the middle of coding up the 2nd method, so it will be a while.
